source/common/network/filter_manager_impl.cc

```diff
@@ -112,6 +112,73 @@ bool FilterManagerImpl::startUpstreamSecureTransport() {
   return false;
 }
 
+void FilterManagerImpl::maybeClose() {
+  if (connection_.state() == Connection::State::Closed) {
+    return;
+  }
+
+  ENVOY_CONN_LOG(trace,
+                 "maybeClose(): remote_close_pending_={}, local_close_pending_={}, "
+                 "write_filter_pending_close_count_={}, read_filter_pending_close_count_={}",
+                 connection_, state_.remote_close_pending_, state_.local_close_pending_,
+                 state_.write_filter_pending_close_count_, state_.read_filter_pending_close_count_);
+
+  // Check if we need to close the connection
+  if ((state_.remote_close_pending_ || state_.local_close_pending_) &&
+      (state_.read_filter_pending_close_count_ == 0 &&
+       state_.write_filter_pending_close_count_ == 0)) {
+    if (latched_close_action_.has_value()) {
+      finalizeClose(latched_close_action_.value());
+    }
+    return;
+  }
+}
+
+void FilterManagerImpl::onConnectionClose(ConnectionCloseAction close_action) {
+  if (connection_.state() == Connection::State::Closed) {
+    return;
+  }
+
+  ASSERT(close_action.isLocalClose() || close_action.isRemoteClose());
+
+  ENVOY_CONN_LOG(trace, "close_action: remote close:{}, local close:{}, close socket:{}",
+                 connection_, close_action.isRemoteClose(), close_action.isLocalClose(),
+                 close_action.closeSocket());
+
+  ENVOY_CONN_LOG(trace,
+                 "onConnectionClose: remote_close_pending_={}, local_close_pending_={}, "
+                 "write_filter_pending_close_count_={}, read_filter_pending_close_count_={}",
+                 connection_, state_.remote_close_pending_, state_.local_close_pending_,
+                 state_.write_filter_pending_close_count_, state_.read_filter_pending_close_count_);
+
+  if (latched_close_action_.has_value() && latched_close_action_->closeSocket()) {
+    if (latched_close_action_->isLocalClose() || !close_action.closeSocket()) {
+      // If the previous close event is a local close and will close the socket, which will only
+      // happen when half close is enabled, or if the current close action does not close the
+      // socket while the previous one does, we keep the previous close action.
+      return;
+    }
+  }
+
+  latched_close_action_ = close_action;
+  state_.local_close_pending_ = close_action.isLocalClose();
+  state_.remote_close_pending_ = close_action.isRemoteClose();
```

Comment:
I think we should only set this after the evaluation the `pending_close_count`.

Reply:
Either way is fine, but move them below the finalized one.